### PR TITLE
Frontend for configuring Maintenance Windows

### PR DESCRIPTION
**Pull request by parlt91**
## What does this PR change?

Frontend part for managing schedules and calendars for the maintenance windows feature

## GUI diff

Before: New Web UI

After:

### Schedules

**Listing Maintenance Schedules:**
ListSchedules

**Creating Schedules:**
CreateSchedule
EditScheduleExpandedDropDown

**Editing Schedules:**
EditScheduleCollapsed
EditSchedule

**Schedule Details:**
ScheduleDetails

**Deleting Schedules:**
DeleteSchedules

### Calendars

**Listing MaintenanceCalendars:**
CalendarList

**Creating Calendars:**
CreateCalendar

**Editing Calendars:**
CalendarEditFile
CalendarEditUrl

**Calendar Details:**
CalendarDetails

**Deleting Calendars:**
CalendarDelete

- [x] **DONE**

## Documentation

Will be provided when the feature is ready to be merged into the master branch

- [x] **DONE**

## Test coverage

Cucumber tests will be provided when the feature is ready to be merged into the master branch

- [x] **DONE**

Fixes: https://github.com/SUSE/spacewalk/issues/11227

- [x] **DONE**

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_lint_checkstyle"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "ruby_rubocop"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

File: java/code/src/com/suse/manager/maintenance/MaintenanceManager.java [hunk 1]
```diff
@@ -246,6 +246,34 @@ protected void save(MaintenanceCalendar calendar) {
             .list();
     }
 
+    /**
+     * List Maintenance Schedules belonging to the given User
+     * @param user the user
+     * @return a list of Maintenance Schedules
+     */
+    @SuppressWarnings("unchecked")
```
Comment (hustodemon) [inline java/code/src/com/suse/manager/maintenance/MaintenanceManager.java:254]: Is this annotation needed?

File: web/html/src/manager/maintenance/maintenance-windows-details.js [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+
+    renderScheduleDetails(data) {
+        return (
+            <BootstrapPanel title={t("Schedule Details")}>
+                <div className="table-responsive">
+                    <table className="table">
+                        <tbody>
+                        <tr>
+                            <td>{t("Schedule Name")}</td>
+                            <td>{t(data.scheduleName)}</td>
```
Comment (ncounter) [inline web/html/src/manager/maintenance/maintenance-windows-details.js:55]: IIRC @mcalmer already caught me in this pattern: we cannot parametrize values inside the `t()` function because texts will be extracted and translated before runtime, do I remember correctly @mcalmer ? If so, we should receive a translated text value from the backend first.
Anyway, I am afraid I am a bit lost in this: is this a field value that the user populate so we should not translate at all or? It would be nice if you can add even some rough and wip screenshot in this PR to get a picture of what we are talking about. Thank you in advance :)

File: web/html/src/manager/maintenance/maintenance-windows-details.js [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+                        </tr>
+                        <tr>
+                            <td>{t("Assigned Calendar")}:</td>
+                            <td>{t(data.calendarName)}</td>
```
Comment (ncounter) [inline web/html/src/manager/maintenance/maintenance-windows-details.js:59]: Same as before

File: web/html/src/manager/maintenance/maintenance-windows-details.js [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+                        </tr>
+                        <tr>
+                            <td>{t("Schedule Type")}:</td>
+                            <td>{t(scheduleTypeToString(data.scheduleType))}</td>
```
Comment (ncounter) [inline web/html/src/manager/maintenance/maintenance-windows-details.js:63]: Same as before

File: web/html/src/manager/maintenance/maintenance-windows-details.js [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+                        </tr>
+                        </tbody>
+                    </table>
+                </div>
+                <DeleteDialog id="delete-modal"
+                              title={t("Delete maintenance schedule")}
+                              content={t("Are you sure you want to delete the selected item? \n" +
+                                  "Deleting the schedule will unassign all systems from this schedule.")}
```
Comment (ncounter) [inline web/html/src/manager/maintenance/maintenance-windows-details.js:71]: It works nice, but I would rephrase just to be less redundant:
`This will remove the current schedule from all the systems assigned to it.`

File: web/html/src/manager/maintenance/maintenance-windows-details.js [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+                        }
+                        </tbody>
+                    </table>
+                </div>
+                <DeleteDialog id="delete-modal"
+                              title={t("Delete maintenance calendar")}
+                              content={
+                                  <Form model={{strategy: this.state.strategy}} onChange={this.setCheck}>
+                                      <div>{t("Are you sure you want to delete the selected item?")}</div>
+                                      <div>{t("Deleting this calendar will unassign all schedules from this calendar.")}</div>
```
Comment (ncounter) [inline web/html/src/manager/maintenance/maintenance-windows-details.js:106]: Same as before

File: web/html/src/manager/maintenance/maintenance-windows-details.js [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+                        <tr>
+                            <td>{t("Used by schedule")}:</td>
+                            <td>{t(data.scheduleNames.map(name => name.name).join(", "))}</td>
+                        </tr>
+                        {data.calendarUrl &&
+                        <tr>
+                            <td>{t("Url")}:</td>
+                            <td>{t(data.calendarUrl)}</td>
+                        </tr>
+                        }
```
Comment (ncounter) [inline web/html/src/manager/maintenance/maintenance-windows-details.js:97]: Code alignment style:
```
{
  data.calendarUrl &&
      <tr>....
      </tr>
}
```

File: web/html/src/manager/maintenance/maintenance-windows-details.js [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+
+        return (
+            <TopPanel
+                title={this.props.data.scheduleName}
+                icon="spacewalk-icon-salt"
+                helpUrl=""
+                button={buttons}
+            >
+                { this.state.messages ?
+                    <Messages items={this.state.messages}/> : null }
```
Comment (ncounter) [inline web/html/src/manager/maintenance/maintenance-windows-details.js:153]: It can stay in a single line I guess. Anyway, isn't the `Messages` component able to digest `items = null`? It would be nice to get rid of the check here.

File: web/html/src/manager/maintenance/maintenance-windows-details.js [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+                        </h4>
+                    </div>
+                    <div className="panel-body">
+                        <pre>
+                            {this.props.data.calendarData}
+                        </pre>
+                    </div>
+                </div>
+                }
+
```
Comment (ncounter) [inline web/html/src/manager/maintenance/maintenance-windows-details.js:170]: Empty line

File: web/html/src/manager/maintenance/maintenance-windows-edit.js [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+            selectedCalendar: 0
+        };
+
+        if(this.isEdit()) {
+            this.setSchedule(this.props.schedule);
+        }
+    }
+
+    setSchedule = (schedule) => {
+        Object.assign(this.state, schedule);
```
Comment (ncounter) [inline web/html/src/manager/maintenance/maintenance-windows-edit.js:34]: `this.state` should be treated as *immutable*. Depending on the data shape maybe you can use something like `this.setState({schedule})`?

File: web/html/src/manager/maintenance/maintenance-windows-edit.js [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+        this.state.calendarDataText = schedule.calendarUrl;
+        Object.assign(this.state, {
```
Comment (ncounter) [inline web/html/src/manager/maintenance/maintenance-windows-edit.js:36]: Same as before

File: web/html/src/manager/maintenance/maintenance-windows-edit.js [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+            calendarAdded: schedule.calendarName ? true : false
+        })
+        if (this.state.type === "schedule" && this.props.calendarNames.length > 1 && this.props.schedule.calendarName) {
+            Object.assign(this.state, {
```
Comment (ncounter) [inline web/html/src/manager/maintenance/maintenance-windows-edit.js:40]: Same as before

File: web/html/src/manager/maintenance/maintenance-windows-edit.js [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+                {(this.isEdit() && !this.state.calendarUrl) &&
+                <Check name="strategy" label={<b>{t("Cancel affected actions")}</b>} divClass="col-md-6 col-md-offset-3" />
+                }
+                <div className="form-horizontal">
+                    <div className="form-group">
+                        <label className="col-md-3 control-label">{t("Calendar data")}:</label>
+                        {(!this.isEdit() || this.state.calendarUrl) &&
+                        <div className={"align-center col-md-" + (this.isEdit() ? "5" : "4")} >
+                            <input type="text" className="form-control"
+                                   style={{textOverflow: "ellipsis"}}
```
Comment (ncounter) [inline web/html/src/manager/maintenance/maintenance-windows-edit.js:191]: Nothing really dangerous but I am against `inline style`. Much better if you use a `className` to maybe just to the one line CSS rule. If it does not exist you can just create a new one in `susemanager-theme.less` maybe.

File: web/html/src/manager/maintenance/maintenance-windows-edit.js [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+                                   placeholder={t("Enter Url to ical file")}
+                                   value={this.state.calendarDataText}
+                                   disabled={this.state.calendarData && !this.state.calendarUrl}
+                                   onChange={this.onCalendarDataTextChanged}/>
+                            {!this.state.calendarData && <b style={{paddingLeft: "25px"}}>or</b>}
```
Comment (ncounter) [inline web/html/src/manager/maintenance/maintenance-windows-edit.js:196]: Same as before for the `inline style`.

Possible hint: if you need just some space between pieces, you can use `&nbsp;`

File: web/html/src/manager/maintenance/maintenance-windows-edit.js [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+                                             icon="fa-refresh"
+                                             target="confirm-modal"
+                                             title={t("Refresh data from url")}
+                                             disabled={this.state.calendarUrl !== this.state.calendarDataText}
+                                />
+                                <DangerDialog id="confirm-modal"
+                                              title="Confirm calendar refresh"
+                                              content={
+                                                  <div>
+                                                      <div>{t("Refreshing the calendar causes affected actions to be rescheduled.")}</div>
```
Comment (ncounter) [inline web/html/src/manager/maintenance/maintenance-windows-edit.js:225]: Rephrase: `Refreshing the calendar will reschedule all the affected actions`

File: web/html/src/manager/maintenance/maintenance-windows-list.js [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+                                onClick={i => this.selectToDelete(i)}
+                            />
+                        </div>
+                    }
+                />
+            </Table>
+            <DeleteDialog id="delete-modal"
+                          title={t("Delete maintenance schedule")}
+                          content={t("Are you sure you want to delete the selected item? \n" +
+                              "Deleting the schedule will unassign all systems from this schedule.")}
```
Comment (ncounter) [inline web/html/src/manager/maintenance/maintenance-windows-list.js:101]: Same as before

File: web/html/src/manager/maintenance/maintenance-windows-list.js [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+                        </div>
+                    }
+                />
+            </Table>
+            <DeleteDialog id="delete-modal"
+                          title={t("Delete maintenance calendar")}
+                          content={
+                              <Form model={{strategy: this.state.strategy}} onChange={this.setCheck}>
+                                  <div>{t("Are you sure you want to delete the selected item?")}</div>
+                                  <div>{t("Deleting this calendar will unassign all schedules from this calendar.")}</div>
```
Comment (ncounter) [inline web/html/src/manager/maintenance/maintenance-windows-list.js:168]: Same as before

File: web/html/src/manager/maintenance/maintenance-windows.css [hunk 18]
```diff
@@ -0,0 +1,14 @@
+:global(.list-tag-link) {
```
Comment (ncounter) [inline web/html/src/manager/maintenance/maintenance-windows.css:1]: Generic comment about style: please use classes defined in `.less` files in the `/branding/css/*` package as much as you can, especially when it comes to colors. As you know the branding can change from time to time and it would be very hard to go through all the customized CSS files in the codebase. Use dedicated CSS files only when it is really necessary for some specific and not reused behavior, otherwise better to create classes and reuse them, so the style of the whole application is consistent. Thanks

File: web/html/src/manager/maintenance/maintenance-windows.js [hunk 19]
```diff
@@ -0,0 +1,249 @@
...
+        return Network.get(endpoint, "application/json").promise
+            .then(item => {
+                this.setState({
+                    selected: item,
+                    action: action
+                });
+                history.pushState(null, null, "#/" + action + "/" +
+                    (this.state.type === "schedule" ? item.scheduleId : item.calendarId));
+            }).catch(this.handleResponseError);
+    }
```
Comment (ncounter) [inline web/html/src/manager/maintenance/maintenance-windows.js:97]: What about to extract the fetching data from the rest of the page? Maybe you can create a `maintenance-windows-api.js` where to collect all the api enpoint calls to fetch data. In case we need to fetch the same data from anywhere else, this could be reused easily.

File: web/html/src/manager/maintenance/maintenance-windows.js [hunk 20]
```diff
@@ -0,0 +1,249 @@
...
+            "application/json"
+        ).promise.then((_) => {
+            const successMsg = <span>{t(
+                (this.state.type === "schedule" ? "Schedule " : "Calendar ") +
+                "successfully" + (this.state.action === "create" ? " created." : " updated."))}</span>
+            const msgs = this.state.messages.concat(MessagesUtils.info(successMsg));
+
+            while (msgs.length > messagesCounterLimit) {
+                msgs.shift();
+            }
```
Comment (ncounter) [inline web/html/src/manager/maintenance/maintenance-windows.js:127]: What about `slicing` the array? https://developer.mozilla.org/it/docs/Web/JavaScript/Reference/Global_Objects/Array/slice

File: web/html/src/manager/maintenance/maintenance-windows.js [hunk 21]
```diff
@@ -0,0 +1,249 @@
...
+            "/rhn/manager/api/maintenance/calendar/refresh",
+            JSON.stringify(item),
+            "application/json"
+        ).promise.then((_) => {
+            const successMsg = <span>{t("Calendar successfully refreshed")}</span>
+            const msgs = this.state.messages.concat(MessagesUtils.info(successMsg));
+
+            while (msgs.length > messagesCounterLimit) {
+                msgs.shift();
+            }
```
Comment (ncounter) [inline web/html/src/manager/maintenance/maintenance-windows.js:174]: Same as before

File: web/html/src/manager/maintenance/maintenance-windows.js [hunk 22]
```diff
@@ -0,0 +1,249 @@
...
+    }
+
+    refreshCalendar(item) {
+        return Network.post(
+            "/rhn/manager/api/maintenance/calendar/refresh",
+            JSON.stringify(item),
+            "application/json"
+        ).promise.then((_) => {
+            const successMsg = <span>{t("Calendar successfully refreshed")}</span>
+            const msgs = this.state.messages.concat(MessagesUtils.info(successMsg));
```
Comment (ncounter) [inline web/html/src/manager/maintenance/maintenance-windows.js:170]: So we are not receiving any data back from the backend? Nothing about the correctness of the action completed?

Anyway, I'd do this in one line only: `MessagesUtils.info(t("Calendar successfully refreshed"))`

Comment (parlt91) [reply web/html/src/manager/maintenance/maintenance-windows.js:170]: Yes we do, the calendar data will be updated. This message will only be displayed in case everything went right. If not we receive an error message from the backend.

File: web/html/src/manager/maintenance/maintenance-windows-details.js (hunk 2 above)
Comment (parlt91) [reply web/html/src/manager/maintenance/maintenance-windows-details.js:55]: Right this value is populated by the user, so it's not translated.

File: web/html/src/manager/maintenance/maintenance-windows-details.js (hunk 3 above)
Comment (parlt91) [reply web/html/src/manager/maintenance/maintenance-windows-details.js:59]: Same as above

File: web/html/src/manager/maintenance/maintenance-windows-details.js (hunk 4 above)
Comment (parlt91) [reply web/html/src/manager/maintenance/maintenance-windows-details.js:63]: I am not sure about this one. Schedule types are `Single` and `Multi`. If those are going to be translated, we need to to something about it.

File: web/html/src/manager/maintenance/maintenance-windows-edit.js (hunk 13 above)
Comment (parlt91) [reply web/html/src/manager/maintenance/maintenance-windows-edit.js:191]: It does definitely not exists int the `.less` file. But this only seems to be used here.
So I would move everything that's unique to this file to it's own css file and the stuff that might
be reusable to `susemanager-theme.less`

File: java/code/src/com/suse/manager/webui/menu/MenuTree.java [hunk 23]
```diff
@@ -288,6 +288,9 @@
                 .addChild(new MenuItem("Action Chains").withPrimaryUrl("/rhn/schedule/ActionChains.do")
                     .withAltUrl("/rhn/schedule/ActionChain.do"))
                 .addChild(new MenuItem("Recurring States").withPrimaryUrl("/rhn/manager/schedule/recurring-states"))
+                .addChild(new MenuItem("Maintenance Windows").withDir("/rhn/manager/schedule/maintenance")
```
[the remark this answers is not in the document]
Comment (parlt91) [reply java/code/src/com/suse/manager/webui/menu/MenuTree.java:291]: I am not sure actually. 'normal' users are not allowed to edit/delete, but I don't know about viewing schedules/calendars. I will clarify this.

File: java/code/src/com/suse/manager/maintenance/MaintenanceManager.java (hunk 1 above)
Comment (parlt91) [reply java/code/src/com/suse/manager/maintenance/MaintenanceManager.java:254]: All the other queries in this file added by mc also seem to have it. And it gets rid of annoying warnings

File: java/code/src/com/suse/manager/webui/controllers/MaintenanceController.java [hunk 24]
```diff
@@ -0,0 +1,498 @@
...
+        }
+        else {
+            Spark.halt(HttpStatus.SC_BAD_REQUEST);
+        }
+        return json(response, ResultJson.success());
+    }
+
+    private static void handleRescheduleResult(List<RescheduleResult> results, String strategy) {
+        results.forEach(result -> {
+            if (!result.isSuccess()) {
```
[the remark this answers is not in the document]
Comment (parlt91) [reply java/code/src/com/suse/manager/webui/controllers/MaintenanceController.java:451]: We definitely want to display all of the failed results, but due to the lack of time I chose the minimalist approach. We cannot use a simple notification message to display multiple failures on the frontend, so we need to come up with something else here.

File: java/code/src/com/suse/manager/webui/controllers/MaintenanceController.java [hunk 25]
```diff
@@ -0,0 +1,498 @@
...
+        else if (json.getCalendarUrl().isBlank() && json.getCalendarData() == null) {
+            Spark.halt(HttpStatus.SC_BAD_REQUEST, GSON.toJson(ResultJson.error(LOCAL.getMessage(
+                    "maintenance.calendar.data.empty"
+            ))));
+        }
+
+        MM.lookupCalendarByUserAndLabel(user, json.getCalendarName()).ifPresentOrElse(
+                /* update existing calendar */
+                calendar -> {
+                    if (json.getCalendarId() == null) {
```
[the remark this answers is not in the document]
Comment (parlt91) [reply java/code/src/com/suse/manager/webui/controllers/MaintenanceController.java:282]: Yes we need it. This case happens, when the user tries to create a new calendar with a name that already exists. Without it you could override an existing calendar by creating a new one with the same name.

[the remark this answers is not in the document]
Comment (hustodemon) [reply java/code/src/com/suse/manager/webui/controllers/MaintenanceController.java:283]: Ok, that makes sense.

File: java/code/src/com/suse/manager/maintenance/MaintenanceManager.java (hunk 1 above)
Comment (hustodemon) [reply java/code/src/com/suse/manager/maintenance/MaintenanceManager.java:372]: What warnings? I don't see any.

Comment (parlt91) [reply java/code/src/com/suse/manager/maintenance/MaintenanceManager.java:372]: It displays a `unchecked assignment` warning if removed.
Anyway I am not opposed to removing it, but we should stay consistent.
Either remove or add it everywhere.

File: java/code/src/com/suse/manager/webui/menu/MenuTree.java (hunk 23 above)
[the remark this answers is not in the document]
Comment (parlt91) [reply java/code/src/com/suse/manager/webui/menu/MenuTree.java:297]: Restricted `destructive` operations to `OrgAdmin` user only as discussed

File: web/html/src/manager/maintenance/maintenance-windows-details.js [hunk 26]
```diff
@@ -38,59 +38,103 @@ class MaintenanceWindowsDetails extends React.Component {
...
+                                  content={t("Are you sure you want to delete the selected item? \n" +
+                                      "This will remove the current schedule from all the systems assigned to it.")}
+                                  onConfirm={() => this.props.onDelete(this.props.data)}
+                    />
+                </BootstrapPanel>
+                {
+                    data.maintenanceWindows !== undefined && data.maintenanceWindows.length > 0 &&
+                    <BootstrapPanel title={t("Upcoming Maintenance Windows")}>
+                        <div className="table-responsive">
+                            <table className="table">
```
Comment (hustodemon) [inline web/html/src/manager/maintenance/maintenance-windows-details.js:72]: Would it be possible to use some existing component for the table (`Table.js`)? Actually, in this PR i see quite some table crafting, maybe it'd make sense to re-use something either. @ncounter What's the general advice here? Re-use components (even for simple tables), or craft tables on-the-fly?

Comment (ncounter) [reply web/html/src/manager/maintenance/maintenance-windows-details.js:72]: I believe I kind of overlooked at this, I remember we discussed with @parlt91 but then we never got back on this. Probably in this case it will be a bit of overengineering work to use the existing component, but we really should. User will never need to use Table.js component feature (ordering by column, comparing values, etc) but it is important to keep every table look&feel fetching structure from the same piece of code.
@parlt91 if you need any help to adapt and make use of the existing table component, just ping me. I think we really should do this now that we have "a bit of time" :wink: 
Thanks @hustodemon for pointing it out

Comment (parlt91) [reply web/html/src/manager/maintenance/maintenance-windows-details.js:72]: Aright. I will look into this. The component seems to work without the headers it's just that there is some empty area remaining where the headers are supposed to be.
I will have to see how it looks. Maybe I need to do some adjustments to the Table but let's see.
